### PR TITLE
Additional alt and control sequences

### LINKED ISSUE #1267: shift + alt + 3 shortcut support
## Details
- OS version: linux|windows 10
- xterm.js version: the one that comes with vs code 1.19.3

run terminal, hit shift + alt + 3
actually alt + 3 passed to terminal
shift + alt + 3 shortcut types # in the start of a line and hits enter.

Actually it is "Alt + #"

Original issue: https://github.com/Microsoft/vscode/issues/42629

### PR DESCRIPTION
**Pull request by saamalik**
* Add punctuation alt sequences (both shift/non-shift versions)
* Add support for ctrl+alt sequences

Follow up to #1225.
Fixes #1267

Sample bash key sequences to try:
- `ctrl+alt+y`: yank first argument from previous command
- `alt+.`: yank last argument previous command

### CODE REVIEW

File: src/Terminal.ts [hunk 1]
```diff
@@ -1713,14 +1713,46 @@ export class Terminal extends EventEmitter implements ITerminal, IInputHandlingT
...
-            result.key = C0.ESC + String.fromCharCode(ev.keyCode + 32);
-          } else if (ev.keyCode === 192) {
-            result.key = C0.ESC + '`';
+            const keyCode = ev.ctrlKey ? ev.keyCode - 64 : ev.keyCode + 32;
+            result.key = C0.ESC + String.fromCharCode(keyCode);
           } else if (ev.keyCode >= 48 && ev.keyCode <= 57) {
             result.key = C0.ESC + (ev.keyCode - 48);
+          } else {
+            const t = (p,s) => !ev.shiftKey ? p : s;
+            switch (ev.keyCode) {
```
Comment (saamalik) [inline src/Terminal.ts]: BTW another possible implementation is having a map structure, e.g:

```
{
   186: [';', ':'],
   187: ['=', '+'],
   188: [',', '<'],
 .....
}
```
And then essentially just do a lookup in map to get actual code.

Let me know if you guys prefer the map instead and I can re-implement.

File: src/Terminal.ts [hunk 2]
```diff
@@ -1713,14 +1713,46 @@ export class Terminal extends EventEmitter implements ITerminal, IInputHandlingT
...
-            result.key = C0.ESC + String.fromCharCode(ev.keyCode + 32);
-          } else if (ev.keyCode === 192) {
-            result.key = C0.ESC + '`';
+            const keyCode = ev.ctrlKey ? ev.keyCode - 64 : ev.keyCode + 32;
+            result.key = C0.ESC + String.fromCharCode(keyCode);
           } else if (ev.keyCode >= 48 && ev.keyCode <= 57) {
             result.key = C0.ESC + (ev.keyCode - 48);
+          } else {
+            const t = (p, s) => !ev.shiftKey ? p : s;
+            switch (ev.keyCode) {
```
Comment (saamalik) [inline src/Terminal.ts]: BTW another possible implementation is having a map structure, e.g:

```
{
   186: [';', ':'],
   187: ['=', '+'],
   188: [',', '<'],
 .....
}
```
And then essentially just do a lookup in map to get actual code.

Let me know if you guys prefer the map instead and I can re-implement.

**Comment by saamalik** [comment]
@Tyriar what are your thoughts on this PR? I don't mind dropping the whole branch and starting afresh if you give me some high-level guidance/pointers.

Ideally we could solve the following two use-cases:
- `ctrl + alt + [a-z]`: these are useful in bash (`ctrl-alt-y`) and also in VIM
- `alt + {special-char}`: useful in bash and again in ViM. 

For the `alt+{special-char}` implementation, I essentially mimicked what _ITerm2_ emits.

**Comment by Tyriar** [comment]
Sorry about not getting to this, a few things:

- 👍 to the map (which should let you remove the `t` function)
- Does ctrl+alt+[a-z] do anything at the moment?
- I'm assuming alt+[special char] only works when `macOptionIsMeta` is true?
- What's up with the tilde key? Can we do nothing there instead?
- How does iTerm2/Terminal.app behave with these new keybindings?

**Comment by saamalik** [comment]
@Tyriar -- cool, will re-implement using the `map` and provide answers to your questions in a bit.

**Comment by saamalik** [comment]
> Does `ctrl+alt+[a-z]` do anything at the moment?

Tried with both `mcOptionIsMeta` _on_ and _off_, but the `ctrl+alt+[a-z]` sequences are completely ignored. Even with `t.setOption('debug', true)` there is _nothing_ printed to the console. 

> `alt+[special char]` only works when macOptionIsMeta is true?

Yup, on Mac OSX both the `alt+[special char]` and `alt+shift+[special char]` only works when macOptionIsMeta; otherwise completely ignored.

Same with `alt+[digit]` (enabled on Win/Linux or when `macOptionIsMeta`). However, `alt+shift+[digit]` also outputs the same sequence as `alt+[digit]`, which is incorrect. As part of the PR I'll also fix the `alt+shift+[digit]` to output the same sequence as iTerm/Terminal.app. With this fix, on bash pressing `alt+shift+*` will expand all matches on command-line. It'll bring the same sequences which are supported on iTerm/Terminal.app.

> What's up with the tilde key? Can we do nothing there instead?

This [tool](https://w3c.github.io/uievents/tools/key-event-viewer.htm) shows the _tilde_ as a _DEADKEY_; however iTerm2 outputs appropriate distinct sequences ``^[` `` and `^[~` for ``alt+` `` and `alt+shift+~`, respectively. PR will add support for both sequences and be consistent with other keys.

> How does iTerm2/Terminal.app behave with these new keybindings?

All of the new mappings are _exactly_ mirroring the sequences emitted with iTerm2/Terminal.app. This was verified using the `ctrl+v` and entering the appropriate key sequence. There might be other keyboard sequences, but at least the ones captured in the map match iTerm2/Terminal.app and other terminals.

**Comment by saamalik** [comment]
@Tyriar this PR is ready for review. Keycode -> key mapping has been re-implemented using a map (actually an `Object`, but let me know if you prefer an actual `Map` even though syntax will look ugly with double, and triple square-brackets).

As I mentioned in the previous comment, all key -> key sequences introduced in this PR match exactly what _iTerm2_ emits. On bash, here are some of the mappings you can try (when `macOptionIsMeta` if on OSX):

    $ touch p1 p2 p3
    $ echo a b c d
    $ <ctrl+alt+y> # a
    $ <alt+.> # d
    $ ls p<alt+shift+/> #list glob-matching files
    $ ls p<alt+shift+8> # insert glob-matching files as arguments

Let me know if you have any questions or would like further changes. Thanks!

**Comment by saamalik** [comment]
@Tyriar friendly ping. What are your thoughts on the PR?

**Comment by Tyriar** [comment]
@saamalik I like it, just might take some time to give it a full review/test. I don't expect many more changes needed to get it merged. I'm hoping to get it in for 3.1.0 😃

**Comment by saamalik** [comment]
@Tyriar btw let me know if you'd like me to squash all commits on top of _origin/master_.

**Comment by Tyriar** [comment]
@saamalik should be fine without that

**Comment by saamalik** [comment]
@Tyriar just another friendly ping to see if we could please get this merged-in?

**Comment by saamalik** [comment]
@Tyriar wohoo -- thanks!

**Pull request merged**